src/control/lib/control/pool.go

```diff
@@ -776,94 +777,73 @@ func (resp *PoolRanksResp) Errors() error {
-	if len(resp.FailedRanks) > 0 {
-		rs := ranklist.RankSetFromRanks(resp.InitialRanks)
-		return errors.Wrapf(err, "pool %s ranks %s", resp.ID, rs.String())
+	rs := ranklist.MustCreateRankSet("")
+	for _, res := range resp.Results {
+		if res.Errored {
+			rs.Add(res.Rank)
+		}
 	}
 
-	rs := ranklist.RankSetFromRanks(resp.FailedRanks)
-	return errors.Wrapf(err, "pool %s ranks %s", resp.ID, rs.String())
-}
-
-// GetResults returns a slice of results from the response and input error.
-func (resp *PoolRanksResp) GetResults(poolID string, reqRanks []ranklist.Rank, errIn error) (results []*PoolRanksResult, err error) {
-	if poolID == "" {
-		return nil, errors.New("pool id is missing")
+	if rs.Count() > 0 {
+		return errors.Errorf("%s %s failed on pool %s",
+			english.PluralWord(rs.Count(), "rank", "ranks"), rs.RangedString(), resp.ID)
 	}
 
-	defer func() {
-		err = errors.Wrapf(err, "pool %s", resp.ID)
-	}()
-
-	if errIn != nil {
-		if len(reqRanks) == 0 {
-			return nil, errors.New("reqRanks is missing")
-		}
+	return nil
+}
 
-		// Return root cause so rank results can be aggregated if required.
-		msgErr := errIn.Error()
-		if f, ok := errors.Cause(errIn).(*fault.Fault); ok {
-			msgErr = f.Error()
-		}
-		results = append(results, &PoolRanksResult{
-			ID:     poolID,
-			Ranks:  reqRanks,
-			Status: int32(daos.MiscError),
-			Msg:    msgErr,
-		})
+type poolRankOpSig func(context.Context, UnaryInvoker, *PoolRanksReq, ranklist.Rank) (*PoolRankResult, error)
 
-		return results, nil
+// Iterate through multiple ranks in PoolRanksReq and call supplied poolRankOp to generate and
+// return a result.
+func getPoolRanksResp(ctx context.Context, rpcClient UnaryInvoker, req *PoolRanksReq, poolRankOp poolRankOpSig) (*PoolRanksResp, error) {
+	if req == nil {
+		return nil, errors.Errorf("nil %T", req)
 	}
-
-	if resp == nil {
-		return nil, errors.Errorf("nil %T", resp)
+	if req.ID == "" {
+		return nil, errors.New("empty pool id")
+	}
+	if len(req.Ranks) == 0 {
+		return nil, errors.New("no ranks in request")
 	}
 
-	if resp.Status != int32(daos.Success) {
-		if len(resp.FailedRanks) == 0 {
-			return nil, errors.New("no failed rank returned with non-zero status")
-		}
-		// Add one result for failed ranks. Note same status applies to all failed ranks.
-		results = append(results, &PoolRanksResult{
-			ID:     poolID,
-			Ranks:  resp.FailedRanks,
-			Status: resp.Status,
-			Msg:    daos.Status(resp.Status).Error(),
-		})
-	} else {
-		if len(resp.FailedRanks) > 0 {
-			return nil, errors.New("failed rank returned with zero status")
-		}
-		if len(resp.SuccessRanks) == 0 {
-			// Expected that at least one result will be generated for each pool.
-			return nil, errors.New("no ranks were operated on")
+	results := []*PoolRankResult{}
+	for _, rank := range req.Ranks {
+		result, err := poolRankOp(ctx, rpcClient, req, rank)
+		if err != nil {
+			return nil, errors.Wrapf(err, "pool %s rank %d", req.ID, rank)
 		}
+		results = append(results, result)
```

Comment:
I think this is a case where we'd want to save the error in the results, so the caller can see if some ranks succeeded?

Reply:
The reason I haven't done that is because most of the failure causes here are  "should not happen" rather than sometimes expected. Happy to change though. Done.